supervisor/process.go

```diff
@@ -0,0 +1,411 @@
+	p.Lock()
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
+	}
+
+	if sig == syscall.SIGTERM {
+		p.setState(Stopping)
+	}
+
+	return syscall.Kill(-p.cmd.Process.Pid, sig)
+}
+
+// StdOut returns the standard output streaming channel
+func (p *Process) StdOut() chan string {
+	return p.streamStdOut
+}
+
+// StdErr returns the standard error streaming channel
+func (p *Process) StdErr() chan string {
+	return p.streamStdErr
+}
+
+// Done returnes a channel that is closed when the process finishes running
+func (p *Process) Done() chan struct{} {
+	return p.done
+}
+
+type StreamOutputOptions func(*StreamOutput)
+
+// ErrLineBufferOverflow is an error type returned when the buffer in StreamOutput used to
+// buffer lines is too small. Try increasing the buffer size if you get this error
+type ErrLineBufferOverflow struct {
+	line       string
+	bufferSize int
+	freeBytes  int
+}
+
+func NewErrLineBufferOverflow(line string, bufsize, freeBytes int) ErrLineBufferOverflow {
+	return ErrLineBufferOverflow{
+		line:       line,
+		bufferSize: bufsize,
+		freeBytes:  freeBytes,
+	}
+}
+
+func (e ErrLineBufferOverflow) Error() string {
+	err := fmt.Sprintf(
+		`line does not contain newline and is %d 
+				bytes too long to buffer (buffer size: %d).
+				Try increasing the buffer size to avoid this error`,
+		len(e.line)-e.bufferSize, e.freeBytes)
+
+	return err
+}
+
+// StreamOutput streams lines of output through the provided channel
+type StreamOutput struct {
+	streamChan chan string
+	bufSize    int
+	buf        []byte
+	lastChar   int
+}
+
+// SetLineBufferSize sets the buffer size of a StreamOutput
+func SetLineBufferSize(size int) StreamOutputOptions {
+	return func(o *StreamOutput) {
+		o.bufSize = size
+		o.buf = make([]byte, size)
+	}
+}
+
+// NewStreamOutput returns an instance of a StreamOuput using the provided channel
+func NewStreamOutput(streamChan chan string, opts ...StreamOutputOptions) *StreamOutput {
+	s := &StreamOutput{
+		streamChan: streamChan,
+		bufSize:    DefaultLineBufferSize,
+		buf:        make([]byte, DefaultLineBufferSize),
+		lastChar:   0,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	return s
+}
+
+// Write implements the io.Writer interface for StreamOutput
+func (s *StreamOutput) Write(b []byte) (int, error) {
+	n := len(b)
+	firstChar := 0
+
+	for {
+		// find the next line in the buffer
+		newLineOffset := bytes.IndexByte(b[firstChar:], '\n')
+
+		if newLineOffset < 0 {
+			break // no newlines in stream
+		}
+
+		lastCharOffset := firstChar + newLineOffset
+
+		// check for carriage return and strip it if present
+		if newLineOffset > 0 && b[lastCharOffset-1] == '\r' {
+			lastCharOffset -= 1
+		}
+
+		var line string
+
+		// preprend any lines if buffer is not empty
+		if s.lastChar > 0 {
+			line = string(s.buf[0:s.lastChar])
+			s.lastChar = 0
+		}
+
+		line += string(b[firstChar:lastCharOffset])
+
+		s.streamChan <- line
```

Comment:
Since you're pushing to the channel on write, how will the clients connecting to the job receive prior logs if the job doesn't produce output anymore?

Reply:
The `Supervisor` holds a buffer to store prior logs for each process. The client can always call `GetOutput()` to get the logs at any time.